src/main/java/org/openlmis/stockmanagement/service/StockCardAggregate.java

```diff
@@ -222,6 +233,30 @@ private long calculateStockoutDays(Map<LocalDate, LocalDate> stockOutDaysMap,
         .sum();
   }
 
+  private int getMostRecentStockOnHandBeforeDate(Map<LocalDate, Integer> stockOnHands,
+                                                 LocalDate date) {
+    if (stockOnHands.isEmpty() || !wasAnyStockOnHandBeforeDate(stockOnHands, date)) {
+      return 0;
+    }
+
+    LocalDate mostRecentDate = stockOnHands.keySet()
+            .stream()
+            .filter(ld -> ld.isBefore(date))
+            .max(Comparator.naturalOrder())
+            .orElseThrow(() -> new ValidationMessageException(ERROR_SOH_NOT_FOUND));
```

Comment:
orElseThrow is a dead code, you already check it in `wasAnyStockOnHandBeforeDate`.
I'm not sure how many items there will be in `stockOnHands`, but you run through it twice in this method - which seems one to much.